db/migrate/20201102160605_split_whodunnit_from_versions.rb

```diff
@@ -0,0 +1,31 @@
+class SplitWhodunnitFromVersions < ActiveRecord::Migration[6.0]
+  def up
+    count = 0
+
+    PaperTrail::Version.all.each do |version|
+      next if version.whodunnit.blank?
+
+      begin
+        Supplier.find(version.whodunnit)
```

Comment:
I reckon it's preferable to have potentially broken references in the data than to execute a query for every single row in the versions table. Looking at the count estimates you can see that there are quite a lot of rows:

```ruby
irb(main):006:0> result =  ActiveRecord::Base.connection.execute("SELECT reltuples as approximate_row_count FROM pg_class WHERE relname = 'versions';")
=> #<PG::Result:0x0000556cd25f8428 status=PGRES_TUPLES_OK ntuples=1 nfields=1 cmd_tuples=1>
irb(main):007:0> result.to_a
=> [{"approximate_row_count"=>5224700.0}]
```

Reply:
I actually think it may be simpler to drop into SQL directly and just let the database do the work 🤔 . You can also only select data that has a known supplier id as per Martyn's query: 

```
# up
UPDATE versions SET supplier_id = whodunit WHERE whodunit IN (SELECT id FROM suppliers)
# down
UPDATE versions SET supplier_id = null WHERE whodunit IN (SELECT id FROM suppliers)
```

Reply:
```ruby
def up
  ActiveRecord::Base.connection.execute(
    'UPDATE versions SET supplier_id = whodunit WHERE whodunit IN (SELECT id FROM suppliers)'
  )
end

def down
  ActiveRecord::Base.connection.execute(
    'UPDATE versions SET supplier_id = null WHERE whodunit IN (SELECT id FROM suppliers)'
  )
end
```

Unfortunately there isn't an index for the whodunnit column so the WHERE clause won't speed us up at all (probably wouldn't anyway given we probably have valid whodunit values for most of the rows.

Reply:
Using this raw SQL, this migration takes 1.7s to run for 100,012 versions, that seems fast enough to me :)

Reply:
Nice :).